File: problems/arrays-and-hashing/242/solve.ts
// npx tsx problems/arrays-and-hashing/242/solve.ts

// Runtime: 7ms, Memory: 57.9MB
function isAnagram(s: string, t: string): boolean {
  if (s.length !== t.length) {
    console.log(false);
    return false;
  }

  const charCount = new Array(26).fill(0);

  for (let i = 0; i < s.length; i++) {
    const index = s[i].charCodeAt(0) - "a".charCodeAt(0);
    charCount[index]++;
  }

  for (let i = 0; i < t.length; i++) {
    const index = t[i].charCodeAt(0) - "a".charCodeAt(0);
    charCount[index]--;
  }

  console.log(charCount.every((count) => count === 0));
  return charCount.every((count) => count === 0);
}

// Runtime: 19ms, Memory: 60.15MB
function isAnagram2(s: string, t: string): boolean {
  if (s.length !== t.length) {
    console.log(false);
    return false;
  }

  const map1 = new Map<string, number>();
  const map2 = new Map<string, number>();

  for (const char of s) {
    map1.set(char, (map1.get(char) || 0) + 1);
  }

  for (const char of t) {
    map2.set(char, (map2.get(char) || 0) + 1);
  }

  for (const [char, count] of map1) {
    if (map2.get(char) !== count) {
      console.log(false);
      return false;
    }
  }

  console.log(true);
  return true;
}

// Runtime: 13ms, Memory: 60.43MB
function isAnagram3(s: string, t: string): boolean {
  if (s.length !== t.length) {
    console.log(false);
    return false;
  }

  const charCount = new Map<string, number>();

  for (const char of s) {
    charCount.set(char, (charCount.get(char) || 0) + 1);
  }

  for (const char of t) {
    const count = charCount.get(char) || 0;
    if (count === 0) {
      console.log(false);
      return false;
    }
    charCount.set(char, count - 1);
  }

  console.log(true);
  return true;
}

console.log("Test 1");
isAnagram3("anagram", "nagaram");

console.log("Test 2");
isAnagram3("rat", "car");

File: problems/arrays-and-hashing/242/solve.go
// go run problems/arrays-and-hashing/242/solve.go

package main

import "fmt"

// Runtime: 2ms, Memory: 4.73MB
func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	count := make(map[rune]int)

	for _, char := range s {
		count[char-'a']++
	}

	for _, char := range t {
		count[char-'a']--
	}

	for _, c := range count {
		if c != 0 {
			fmt.Println(false)
			return false
		}
	}

	fmt.Println(true)
	return true
}

// Runtime: 4ms, Memory: 4.69MB
func isAnagram2(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	map1 := make(map[rune]int)
	map2 := make(map[rune]int)

	for _, char := range s {
		map1[char]++
	}

	for _, char := range t {
		map2[char]++
	}

	for char, count := range map1 {
		if map2[char] != count {
			fmt.Println(false)
			return false
		}
	}

	fmt.Println(true)
	return true
}

// Runtime: 0ms, Memory: 4.81MB
func isAnagram3(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	charCount := [26]int{}

	for _, char := range s {
		charCount[char-'a']++
	}

	for _, char := range t {
		charCount[char-'a']--
	}

	for _, count := range charCount {
		if count != 0 {
			fmt.Println(false)
			return false
		}
	}

	fmt.Println(true)
	return true
}

func main() {
	fmt.Printf("Test 1\n")
	isAnagram3("anagram", "nagaram")

	fmt.Printf("Test 2\n")
	isAnagram3("rat", "car")
}
